Guide for creating alert / action types in the UI (#55963)
elastic/kibana: x-pack/plugins/triggers_actions_ui/README.md

    - [Create and register new alert type UI example](#create-and-register-new-alert-type-ui-example)
    - [Common expression components](#common-expression-components)
      - [WHEN expression component](#when-expression-component)
      - [OF expression component](#of-expression-component)
      - [GROUPED BY expression component](#grouped-by-expression-component)
      - [FOR THE LAST expression component](#for-the-last-expression-component)
      - [THRESHOLD expression component](#threshold-expression-component)
    - [Embed the Create Alert flyout within any Kibana plugin](#embed-the-create-alert-flyout-within-any-kibana-plugin)
  - [Build and register Action Types](#build-and-register-action-types)
    - [Built-in Action Types](#built-in-action-types)
      - [Server log](#server-log)
      - [Email](#email)
      - [Slack](#slack)
      - [Index](#index)
      - [Webhook](#webhook)
      - [PagerDuty](#pagerduty)
    - [Action type model definition](#action-type-model-definition)
    - [Register action type model](#register-action-type-model)
    - [Create and register new action type UI example](#reate-and-register-new-action-type-ui-example)

## Built-in Alert Types

> Kibana ships with several built-in alert types:

Author: Yes, currently we have only one, but planning to have a more for the future releases

```
|Property|Description|
|---|---|
|id|Alert type id. Should be the same as on the server side.|
|name|Name of alert type, that will be displayed on the select card in UI.|
|iconClass|Icon of alert type, that will be displayed on the select card in UI.|
|validate|Validation function for the alert params.|
|alertParamsExpression|React functional component for building UI of the current alert type params.|
|defaultActionMessage|Optional property for providing default message for all added actions with `message` property.|

IMPORTANT! The current UI support only a single action group. 
Action groups is mapped from server API result for [GET /api/alert/types: List alert types](https://github.com/elastic/kibana/tree/master/x-pack/legacy/plugins/alerting#get-apialerttypes-list-alert-types).
Server side alert type model:
```
export interface AlertType {
  id: string;
  name: string;
  validate?: {
    params?: { validate: (object: any) => any };
  };
  actionGroups: string[];
  executor: ({ services, params, state }: AlertExecutorOptions) => Promise<State | void>;
}
```
Only the default (which means first item of the array) action group is displayed in the current UI.

> Design of user interface and server API for multiple action groups is on the stage of discussion and development.

Author: Yes, I want to say "under discussion and development"

## Register alert type model

There are two ways of registering a new alert type:

> 1. Directly in `triggers_actions_ui` plugin. In this case alert type will be available in Create Alert flyout of the Alerts and Actions management section.

Reviewer: ```suggestion
1. Directly in the `triggers_actions_ui` plugin. In this case, the alert type will be available in the Create Alert flyout of the Alerts and Actions management section.
```

Registration code for a new alert type model should be added to the file `x-pack/plugins/triggers_actions_ui/public/application/components/builtin_alert_types/index.ts`
Only registered alert types are available in UI.

> 2. Register alert type in other plugin. In this case alert type will be available only in current plugin UI. 

Reviewer: ```suggestion
2. Register the alert type in another plugin. In this case, the alert type will be available only in the current plugin UI. 
```

It should be done by importing dependency `TriggersAndActionsUIPublicPluginSetup` and adding the next code on plugin setup:

```
function getSomeNewAlertType() {
  return { ... } as AlertTypeModel;
}

triggers_actions_ui.alertTypeRegistry.register(getSomeNewAlertType());
```

## Create and register new alert type UI example

Before registering a UI for a new Alert Type, you should first register the type on the server-side by following the Alerting guide: https://github.com/elastic/kibana/tree/master/x-pack/legacy/plugins/alerting#example 

Alert type UI is expected to be defined as `AlertTypeModel` object.

Below is a list of steps that should be done to build and register a new alert type with the name `Example Alert Type`:

> 1. At any suitable place in Kibana create a file, which will expose an object implementing interface [AlertTypeModel](https://github.com/elastic/kibana/blob/55b7905fb5265b73806006e7265739545d7521d0/x-pack/legacy/plugins/triggers_actions_ui/np_ready/public/types.ts#L83). Example:

Reviewer: ```suggestion
1. At any suitable place in Kibana, create a file, which will expose an object implementing interface [AlertTypeModel](https://github.com/elastic/kibana/blob/55b7905fb5265b73806006e7265739545d7521d0/x-pack/legacy/plugins/triggers_actions_ui/np_ready/public/types.ts#L83). Example:
```

```
import { AlertTypeModel } from '../../../../types';
import { ExampleExpression } from './expression';
import { validateExampleAlertType } from './validation';

export function getAlertType(): AlertTypeModel {
  return {
    id: 'example',
    name: 'Example Alert Type',
    iconClass: 'bell',
    alertParamsExpression: ExampleExpression,
    validate: validateExampleAlertType,
    defaultActionMessage: 'Alert [{{ctx.metadata.name}}] has exceeded the threshold',
  };
}
```

> Fields of this object `AlertTypeModel` will be mapped properly in UI below.

Reviewer: ```suggestion
Fields of this object `AlertTypeModel` will be mapped properly in the UI below.
```

  return (
    <Fragment>
      <EuiFlexGroup gutterSize="s" wrap>
        <EuiFlexItem grow={false}>
          <WhenExpression
            aggType={testAggType ?? 'count'} // defult is 'count'
            onChangeSelectedAggType={(selectedAggType: string) => {
              console.log(`Set alert type params field "aggType" value as ${selectedAggType}`);
              setAggType(selectedAggType);
            }}
          />
        </EuiFlexItem>
        {aggType && builtInAggregationTypes[aggType].fieldRequired ? (
          <EuiFlexItem grow={false}>
            <OfExpression
              aggField={testAggField}
              fields={[{ normalizedType: 'number', name: 'test' }]} // can be some data from server API
              aggType={aggType}
              errors={errors}
              onChangeSelectedAggField={(selectedAggField?: string) =>
                console.log(`Set alert type params field "aggField" value as ${selectedAggField}`)
              }
            />
          </EuiFlexItem>
        ) : null}
      </EuiFlexGroup>
    </Fragment>
  );
};

```
This alert type form becomes available, when the card of `Example Alert Type` is selected.

> Each expression word here is `EuiExpression` component and implement the basic aggregation, grouping and comparison methods.

Reviewer: ```suggestion
Each expression word here is `EuiExpression` component and implements the basic aggregation, grouping and comparison methods.
```

> Expression components, which can be embed to different alert types are described here [Common expression components](#common-expression-components).

Reviewer: ```suggestion
Expression components, which can be embedded to different alert types, are described here [Common expression components](#common-expression-components).
```

3. Define alert type params validation using the property of `AlertTypeModel` `validate`: 
```
import { i18n } from '@kbn/i18n';
import { ValidationResult } from '../../../../types';

export function validateExampleAlertType({
  testAggField,
}: {
  testAggField: string;
}): ValidationResult {
  const validationResult = { errors: {} };
  const errors = {
    aggField: new Array<string>(),
  };
  validationResult.errors = errors;
  if (!testAggField) {
    errors.aggField.push(
      i18n.translate('xpack.triggersActionsUI.components.example.error.requiredTestAggFieldText', {
        defaultMessage: 'Test aggregation field is required.',
      })
    );
  }
  return validationResult;
}
```

4. Extend registration code with the new alert type register in the file `x-pack/plugins/triggers_actions_ui/public/application/components/builtin_alert_types/index.ts`
```
import { getAlertType as getExampledAlertType } from './example';
...

...
alertTypeRegistry.register(getExampledAlertType());
```

After these four steps, the new `Example Alert Type` is available in UI of Create flyout:
![Example Alert Type is in the select cards list](https://i.imgur.com/j71AEQV.png)

> Click on select card for `Example Alert Type` will open expression form that was created in step 2:

Reviewer: ```suggestion
Click on the select card for `Example Alert Type` to open the expression form that was created in step 2:
```

![Example Alert Type expression with validation](https://i.imgur.com/Z0jIwCS.png)

## Common expression components

### WHEN expression component

![WHEN](https://i.imgur.com/7bYlxXK.png)

```
<WhenExpression
  aggType={aggType ?? DEFAULT_VALUES.AGGREGATION_TYPE}
  onChangeSelectedAggType={(selectedAggType: string) =>
    setAlertParams('aggType', selectedAggType)
  }
/>
```

Props definition:
```
interface WhenExpressionProps {
  aggType: string;
  customAggTypesOptions?: { [key: string]: AggregationType };
  onChangeSelectedAggType: (selectedAggType: string) => void;
  popupPosition?: 'upCenter' | 'upLeft' | 'upRight' | 'downCenter' | 'downLeft'
    | 'downRight' | 'leftCenter' | 'leftUp' | 'leftDown' | 'rightCenter' | 'rightUp' | 'rightDown';
}
```

|Property|Description|
|---|---|
|aggType|Selected aggregation type that will be set as the alert type property.|
|customAggTypesOptions|(Optional) List of aggregation types that replaces the default options defined in constants `x-pack/plugins/triggers_actions_ui/public/common/constants/aggregation_types.ts`.|

> |onChangeSelectedAggType|event handler that will be excuted selected aggregation type is changed.|

Reviewer: ```suggestion
|onChangeSelectedAggType|event handler that will be executed when selected aggregation type is changed.|
```

> |popupPosition|(Optional) expression popup position. Default is `downLeft`, recomemded to change it for a small parent window space.|

Reviewer: ```suggestion
|popupPosition|(Optional) expression popup position. Default is `downLeft`.  Recommend changing it for a small parent window space.|
```

```
<OfExpression
  aggField={aggField}
  fields={esFields}
  aggType={aggType}
  errors={errors}
  onChangeSelectedAggField={(selectedAggField?: string) =>
    setAlertParams('aggField', selectedAggField)
  }
/>
```

Props definition:
```
interface OfExpressionProps {
  aggType: string;
  aggField?: string;
  errors: { [key: string]: string[] };
  onChangeSelectedAggField: (selectedAggType?: string) => void;
  fields: Record<string, any>;
  customAggTypesOptions?: {
    [key: string]: AggregationType;
  };
  popupPosition?: 'upCenter' | 'upLeft' | 'upRight' | 'downCenter' | 'downLeft'
    | 'downRight' | 'leftCenter' | 'leftUp' | 'leftDown' | 'rightCenter' | 'rightUp' | 'rightDown';
}
```

|Property|Description|
|---|---|
|aggType|Selected aggregation type that will be set as the alert type property.|
|aggField|Selected aggregation field that will be set as the alert type property.|
|errors|List of errors with proper messages for the alert params that should be validated. In current component is validated `aggField`.|
|onChangeSelectedAggField|Event handler that will be excuted if selected aggregation field is changed.|

> |fields|Fields list that will be available in the OF `Select a field` dropdow.|

Reviewer: ```suggestion
|fields|Fields list that will be available in the OF `Select a field` dropdown.|
```

|customAggTypesOptions|(Optional) List of aggregation types that replaces the default options defined in constants `x-pack/plugins/triggers_actions_ui/public/common/constants/aggregation_types.ts`.|

> |popupPosition|(Optional) expression popup position. Default is `downRight`, recomemded to change it for a small parent window space.|

Reviewer: ```suggestion
|popupPosition|(Optional) expression popup position. Default is `downRight`. Recommend changing it for a small parent window space.|
```

  onChangeSelectedTermSize: (selectedTermSize?: number) => void;
  onChangeSelectedTermField: (selectedTermField?: string) => void;
  onChangeSelectedGroupBy: (selectedGroupBy?: string) => void;
  fields: Record<string, any>;
  customGroupByTypes?: {
    [key: string]: GroupByType;
  };
  popupPosition?: 'upCenter' | 'upLeft' | 'upRight' | 'downCenter' | 'downLeft'
    | 'downRight' | 'leftCenter' | 'leftUp' | 'leftDown' | 'rightCenter' | 'rightUp' | 'rightDown';
}
```

|Property|Description|
|---|---|
|groupBy|Selected group by type that will be set as the alert type property.|
|termSize|Selected term size that will be set as the alert type property.|
|termField|Selected term field that will be set as the alert type property.|
|errors|List of errors with proper messages for the alert params that should be validated. In current component is validated `termSize` and `termField`.|
|onChangeSelectedTermSize|Event handler that will be excuted if selected term size is changed.|
|onChangeSelectedTermField|Event handler that will be excuted if selected term field is changed.|
|onChangeSelectedGroupBy|Event handler that will be excuted if selected group by is changed.|

> |fields|Fields list that will an options for the `termField` dropdow.|

Reviewer: ```suggestion
|fields|Fields list with options for the `termField` dropdown.|
```

|customGroupByTypes|(Optional) List of group by types that replaces the default options defined in constants `x-pack/plugins/triggers_actions_ui/public/common/constants/group_by_types.ts`.|

> |popupPosition|(Optional) expression popup position. Default is `downLeft`, recomemded to change it for a small parent window space.|

Reviewer: ```suggestion
|popupPosition|(Optional) expression popup position. Default is `downLeft`. Recommend changing it for a small parent window space.|
```

    setAlertParams('timeWindowSize', selectedWindowSize)
  }
  onChangeWindowUnit={(selectedWindowUnit: any) =>
    setAlertParams('timeWindowUnit', selectedWindowUnit)
  }
/>
```

Props definition:
```
interface ForLastExpressionProps {
  timeWindowSize?: number;
  timeWindowUnit?: string;
  errors: { [key: string]: string[] };
  onChangeWindowSize: (selectedWindowSize: number | '') => void;
  onChangeWindowUnit: (selectedWindowUnit: string) => void;
  popupPosition?: 'upCenter' | 'upLeft' | 'upRight' | 'downCenter' | 'downLeft'
    | 'downRight' | 'leftCenter' | 'leftUp' | 'leftDown' | 'rightCenter' | 'rightUp' | 'rightDown';
}
```

|Property|Description|
|---|---|
|timeWindowSize|Selected time window size that will be set as the alert type property.|
|timeWindowUnit|Selected time window unit that will be set as the alert type property.|
|errors|List of errors with proper messages for the alert params that should be validated. In current component is validated `termWindowSize`.|
|onChangeWindowSize|Event handler that will be excuted if selected window size is changed.|
|onChangeWindowUnit|Event handler that will be excuted if selected window unit is changed.|

> |popupPosition|(Optional) expression popup position. Default is `downLeft`, recomemded to change it for a small parent window space.|

Reviewer: ```suggestion
|popupPosition|(Optional) expression popup position. Default is `downLeft`. Recommend changing it for a small parent window space.|
```

  onChangeSelectedThresholdComparator: (selectedThresholdComparator?: string) => void;
  onChangeSelectedThreshold: (selectedThreshold?: number[]) => void;
  customComparators?: {
    [key: string]: Comparator;
  };
  threshold?: number[];
  popupPosition?: 'upCenter' | 'upLeft' | 'upRight' | 'downCenter' | 'downLeft'
    | 'downRight' | 'leftCenter' | 'leftUp' | 'leftDown' | 'rightCenter' | 'rightUp' | 'rightDown';
}
```

|Property|Description|
|---|---|
|thresholdComparator|Selected time window size that will be set as the alert type property.|
|threshold|Selected time window size that will be set as the alert type property.|
|errors|List of errors with proper messages for the alert params that should be validated. In current component is validated `threshold0` and `threshold1`.|
|onChangeSelectedThresholdComparator|Event handler that will be excuted if selected threshold comparator is changed.|
|onChangeSelectedThreshold|Event handler that will be excuted if selected threshold is changed.|
|customComparators|(Optional) List of comparators that replaces the default options defined in constants `x-pack/plugins/triggers_actions_ui/public/common/constants/comparators.ts`.|

> |popupPosition|(Optional) expression popup position. Default is `downLeft`, recomemded to change it for a small parent window space.|

Reviewer: ```suggestion
|popupPosition|(Optional) expression popup position. Default is `downLeft`. Recommend changing it for a small parent window space.|
```

2. Add Create Alert flyout to React component:
```
// import section
import { AlertsContextProvider, AlertAdd } from '../../../../../../../triggers_actions_ui/public';

// in the component state definition section
const [alertFlyoutVisible, setAlertFlyoutVisibility] = useState<boolean>(false);

// UI control item for open flyout
<EuiButton
  fill
  iconType="plusInCircle"
  iconSide="left"
  onClick={() => setAlertFlyoutVisibility(true)}
>
  <FormattedMessage
    id="emptyButton"
    defaultMessage="Create alert"
  />
</EuiButton>

// in render section of component
<AlertsContextProvider
  value={{
    addFlyoutVisible: alertFlyoutVisible,
    setAddFlyoutVisibility: setAlertFlyoutVisibility,
    http,
    actionTypeRegistry: triggers_actions_ui.actionTypeRegistry,
    alertTypeRegistry: triggers_actions_ui.alertTypeRegistry,
    toastNotifications: toasts,
    uiSettings,
    charts,
    dataFieldsFormats,
  }}
>
  <AlertAdd consumer={'watcher'}  />
</AlertsContextProvider>
```

AlertAdd Props definition:
```
interface AlertAddProps {
  consumer: string;
  alertTypeId?: string;
  canChangeTrigger?: boolean;
}
```

|Property|Description|
|---|---|

> |consumer|Name of the plugin, which creating an alert.|

Reviewer: ```suggestion
|consumer|Name of the plugin that creates an alert.|
```

AlertsContextProvider value options:
```
export interface AlertsContextValue {
  addFlyoutVisible: boolean;
  setAddFlyoutVisibility: React.Dispatch<React.SetStateAction<boolean>>;
  reloadAlerts?: () => Promise<void>;
  http: HttpSetup;
  alertTypeRegistry: TypeRegistry<AlertTypeModel>;
  actionTypeRegistry: TypeRegistry<ActionTypeModel>;
  uiSettings?: IUiSettingsClient;
  toastNotifications?: Pick<
    ToastsApi,
    'get$' | 'add' | 'remove' | 'addSuccess' | 'addWarning' | 'addDanger' | 'addError'
  >;
  charts?: ChartsPluginSetup;
  dataFieldsFormats?: Pick<FieldFormatsRegistry, 'register'>;
}
```

|Property|Description|
|---|---|
|addFlyoutVisible|Visibility state of the Create Alert flyout.|
|setAddFlyoutVisibility|Function for changing visibility state of the Create Alert flyout.|
|reloadAlerts|Optional function, which will be executed if alert was saved sucsessfuly.|
|http|HttpSetup needed for executing API calls.|
|alertTypeRegistry|Registry for alert types.|
|actionTypeRegistry|Registry for action types.|
|uiSettings|Optional property, which is needed to display visualization of alert type expression. Will be changed after visualization refactoring.|

> |toastNotifications|Optional toasts messages.|

Reviewer: ```suggestion
|toastNotifications|Optional toast messages.|
```

|charts|Optional property, which is needed to display visualization of alert type expression. Will be changed after visualization refactoring.|
|dataFieldsFormats|Optional property, which is needed to display visualization of alert type expression. Will be changed after visualization refactoring.|

## Build and register Action Types

Kibana ships with a set of built-in action types UI:

|Type|Id|Description|
|---|---|---|
|[Server log](#server-log)|`.log`|Logs messages to the Kibana log using `server.log()`|
|[Email](#email)|`.email`|Sends an email using SMTP|

> |[Slack](#slack)|`.slack`|Posts a message to a slack channel|

Reviewer: ```suggestion
|[Slack](#slack)|`.slack`|Posts a message to a Slack channel|
```

|[Index](#index)|`.index`|Indexes document(s) into Elasticsearch|

> |[Webhook](#webhook)|`.webhook`|Send a payload to a web service using HTTP POST or PUT|

Reviewer: ```suggestion
|[Webhook](#webhook)|`.webhook`|Sends a payload to a web service using HTTP POST or PUT|
```

> |[PagerDuty](#pagerduty)|`.pagerduty`|Trigger, resolve, or acknowlege an incident to a PagerDuty service|

Reviewer: ```suggestion
|[PagerDuty](#pagerduty)|`.pagerduty`|Triggers, resolves, or acknowledges an incident to a PagerDuty service|
```

Every action type should be registered server side, and can be optionally registered client side. 

> Only action types registered on both - client and server will be displayed in Alerts and Actions UI.

Reviewer: ```suggestion
Only action types registered on both client and server will be displayed in the Alerts and Actions UI.
```

```
export function getActionType(): ActionTypeModel {
  return {
    id: '.pagerduty',
    iconClass: 'apps',
    selectMessage: i18n.translate(
      'xpack.triggersActionsUI.components.builtinActionTypes.pagerDutyAction.selectMessageText',
      {
        defaultMessage: 'Send an event in PagerDuty.',
      }
    ),
    actionTypeTitle: i18n.translate(
      'xpack.triggersActionsUI.components.builtinActionTypes.pagerDutyAction.actionTypeTitle',
      {
        defaultMessage: 'Send to PagerDuty',
      }
    ),
    validateConnector: (action: PagerDutyActionConnector): ValidationResult => {
      // validation of connector properties implementation
    },
    validateParams: (actionParams: PagerDutyActionParams): ValidationResult => {
      // validation of action params implementation
    },
    actionConnectorFields: PagerDutyActionConnectorFields,
    actionParamsFields: PagerDutyParamsFields,
  };
}
```

![PagerDuty connector card](https://i.imgur.com/Br8MuKG.png)

![PagerDuty connector form](https://i.imgur.com/DZpCfRv.png)

and action params form available in Create Alert form:
![PagerDuty action form](https://i.imgur.com/xxXmhMK.png)

## Action type model definition

> Each action type should be defined as `ActionTypeModel` object with the next properties:

Reviewer: ```suggestion
Each action type should be defined as an `ActionTypeModel` object with the following properties:
```

```
  id: string;
  iconClass: string;
  selectMessage: string;
  actionTypeTitle?: string;
  validateConnector: (connector: any) => ValidationResult;
  validateParams: (actionParams: any) => ValidationResult;
  actionConnectorFields: React.FunctionComponent<any> | null;
  actionParamsFields: any;
```
|Property|Description|
|---|---|
|id|Action type id. Should be the same as on server side.|
|iconClass|Icon of action type, that will be displayed on the select card in UI.|
|selectMessage|Short description of action type responsibility, that will be displayed on the select card in UI.|
|validateConnector|Validation function for action connector.|
|validateParams|Validation function for action params.|
|actionConnectorFields|React functional component for building UI of current action type connector.|
|actionParamsFields|React functional component for building UI of current action type params. Displayed as a part of Create Alert flyout.|

## Register action type model

There are two ways to register a new action type UI:

> 1. Directly in `triggers_actions_ui` plugin. In this case action type will be available in the Alerts and Actions management section.

Reviewer: ```suggestion
1. Directly in `triggers_actions_ui` plugin. In this case, the action type will be available in the Alerts and Actions management section.
```

Registration code for a new action type model should be added to the file `x-pack/plugins/triggers_actions_ui/public/application/components/builtin_action_types/index.ts`
Only registered action types are available in UI.

> 2. Register action type in other plugin. In this case action type will be available only in the current plugin UI. 

Reviewer: ```suggestion
2. Register action type in another plugin. In this case, the action type will be available only in the current plugin UI. 
```

It should be done by importing dependency `TriggersAndActionsUIPublicPluginSetup` and adding the next code on plugin setup:

```
function getSomeNewActionType() {
  return { ... } as ActionTypeModel;
}

triggers_actions_ui.actionTypeRegistry.register(getSomeNewActionType());
```

## Create and register new action type UI

Before starting the UI implementation, the [server side registration](https://github.com/elastic/kibana/tree/master/x-pack/plugins/actions#kibana-actions-configuration) should be done first.

Action type UI is expected to be defined as `ActionTypeModel` object.

Below is a list of steps that should be done to build and register a new action type with the name `Example Action Type`:

> 1. At any suitable place in Kibana create a file, which will expose an object implementing interface [ActionTypeModel]:

Reviewer: ```suggestion
1. At any suitable place in Kibana, create a file, which will expose an object implementing interface [ActionTypeModel]:
```

- Create and register new action type.
- Embed Create Alert flyout to Kibana plugins.

To enable Alerts and Actions UI functionality, the next configuration settings should be provided:
```
xpack.alerting.enabled: true
xpack.actions.enabled: true
xpack.triggers_actions_ui.enabled: true
xpack.triggers_actions_ui.createAlertUiEnabled: true
```

-----

Table of Contents

- [Kibana Alerts and Actions UI](#kibana-alerts-and-actions-ui)
  - [Build and register Alert Types](#build-and-register-alert-types)
    - [Built-in Alert Types](#built-in-alert-types)
      - [Index Threshold Alert](#index-threshold-alert)
    - [Alert type model definition](#alert-type-model-definition)
    - [Register alert type model](#register-alert-type-model)
    - [Create and register new alert type UI example](#create-and-register-new-alert-type-ui-example)
    - [Common expression components](#common-expression-components)
      - [WHEN expression component](#when-expression-component)
      - [OF expression component](#of-expression-component)
      - [GROUPED BY expression component](#grouped-by-expression-component)
      - [FOR THE LAST expression component](#for-the-last-expression-component)

>       - [THRESHOLD expression component](#threshold-expression-component)

Reviewer 2: I'm wondering if maybe we should extract the Expression types into a separate guide?
They feel more generic that the Alerting UI, no?

  };
  actionGroups: string[];
  executor: ({ services, params, state }: AlertExecutorOptions) => Promise<State | void>;
}
```
Only one default (which means first item of the array) action group is displayed in the current UI.
Design of user interface and server API for multiple action groups is on the stage of discussion and development.

## Register alert type model

There are two ways of registration new alert type:

1. Directly in `triggers_actions_ui` plugin. In this case alert type will be available in Create Alert flyout of the Alerts and Actions management section.
Registration code for a new alert type model should be added to the file `x-pack/plugins/triggers_actions_ui/public/application/components/builtin_alert_types/index.ts`
Only registered alert types are available in UI.

2. Register alert type in other plugin. In this case alert type will be available only in current plugin UI. 
It should be done by importing dependency `TriggersAndActionsUIPublicPluginSetup` and adding the next code on plugin setup:

```
function getSomeNewAlertType() {
  return { ... } as AlertTypeModel;
}

triggers_actions_ui.alertTypeRegistry.register(getSomeNewAlertType());
```

> 

Reviewer 2: It might be worth linking from here to the New Platform guide that explains how to define plugin dependencies.

## Create and register new alert type UI example

> To be able to add UI for a new Alert type the proper server API https://github.com/elastic/kibana/tree/master/x-pack/legacy/plugins/alerting#example recommended to be done first.

Reviewer 2: ```suggestion
Before registering a UI for a new Alert Type, you should first register the type on the server-side by following the Alerting guide: https://github.com/elastic/kibana/tree/master/x-pack/legacy/plugins/alerting#example 
```

> To be able to add UI for a new Alert type the proper server API https://github.com/elastic/kibana/tree/master/x-pack/legacy/plugins/alerting#example recommended to be done first.

Reviewer 3: ```suggestion
To be able to add UI for a new Alert type, first implement the server API [example implementation](https://github.com/elastic/kibana/tree/master/x-pack/legacy/plugins/alerting#example).
```

          <WhenExpression
            aggType={testAggType ?? 'count'} // defult is 'count'
            onChangeSelectedAggType={(selectedAggType: string) => {
              console.log(`Set alert type params field "aggType" value as ${selectedAggType}`);
              setAggType(selectedAggType);
            }}
          />
        </EuiFlexItem>
        {aggType && builtInAggregationTypes[aggType].fieldRequired ? (
          <EuiFlexItem grow={false}>
            <OfExpression
              aggField={testAggField}
              fields={[{ normalizedType: 'number', name: 'test' }]} // can be some data from server API
              aggType={aggType}
              errors={errors}
              onChangeSelectedAggField={(selectedAggField?: string) =>
                console.log(`Set alert type params field "aggField" value as ${selectedAggField}`)
              }
            />
          </EuiFlexItem>
        ) : null}
      </EuiFlexGroup>
    </Fragment>
  );
};

```
This alert type form becomes available, when the card of `Example Alert Type` is selected.
Each expression word here is `EuiExpression` component and implement the basic aggregation, grouping and comparison methods.

> Expression components, which can be embed to different alert types is described here [Common expression components](#common-expression-components).

Reviewer 3: ```suggestion
Expression components, which can be embed to different alert types are described here [Common expression components](#common-expression-components).
```

3. Define alert type params validation using the property of `AlertTypeModel` `validate`: 
```
import { i18n } from '@kbn/i18n';
import { ValidationResult } from '../../../../types';

export function validateExampleAlertType({
  testAggField,
}: {
  testAggField: string;
}): ValidationResult {
  const validationResult = { errors: {} };
  const errors = {
    aggField: new Array<string>(),
  };
  validationResult.errors = errors;
  if (!testAggField) {
    errors.aggField.push(
      i18n.translate('xpack.triggersActionsUI.components.example.error.requiredTestAggFieldText', {
        defaultMessage: 'Test aggregation field is required.',
      })
    );
  }
  return validationResult;
}
```

4. Extend registration code with the new alert type register in the file `x-pack/plugins/triggers_actions_ui/public/application/components/builtin_alert_types/index.ts`
```
import { getAlertType as getExampledAlertType } from './example';
...

...
alertTypeRegistry.register(getExampledAlertType());
```

> After this four steps new `Example Alert Type` is available in UI of Create flyout:

Reviewer 3: ```suggestion
After these four steps, the new `Example Alert Type` is available in UI of Create flyout:
```

![Example Alert Type is in the select cards list](https://i.imgur.com/j71AEQV.png)

> Click on select cart for `Example Alert Type` will open expression form that was created in step 2:

Reviewer 2: ```suggestion
Click on select card for `Example Alert Type` will open expression form that was created in step 2:
```

    [key: string]: Comparator;
  };
  threshold?: number[];
  popupPosition?: 'upCenter' | 'upLeft' | 'upRight' | 'downCenter' | 'downLeft'
    | 'downRight' | 'leftCenter' | 'leftUp' | 'leftDown' | 'rightCenter' | 'rightUp' | 'rightDown';
}
```

|Property|Description|
|---|---|
|thresholdComparator|Selected time window size that will be set as the alert type property.|
|threshold|Selected time window size that will be set as the alert type property.|
|errors|List of errors with proper messages for the alert params that should be validated. In current component is validated `threshold0` and `threshold1`.|
|onChangeSelectedThresholdComparator|Event handler that will be excuted if selected threshold comparator is changed.|
|onChangeSelectedThreshold|Event handler that will be excuted if selected threshold is changed.|
|customComparators|(Optional) List of comparators that replaces the default options defined in constants `x-pack/plugins/triggers_actions_ui/public/common/constants/comparators.ts`.|
|popupPosition|(Optional) expression popup position. Default is `downLeft`, recomemded to change it for a small parent window space.|

## Embed Create Alert flyout to Kibana plugins

> To embed Create Alert flyout to any place in Kibana the next code should be specified under the React component file:

Reviewer 2: ```suggestion
Follow the instructions bellow to embed the Create Alert flyout within any Kibana plugin:
```

|alertTypeRegistry|Registry for alert types.|
|actionTypeRegistry|Registry for action types.|
|uiSettings|Optional property, which is needed to display visualization of alert type expression. Will be changed after visualization refactoring.|
|toastNotifications|Optional toasts messages.|
|charts|Optional property, which is needed to display visualization of alert type expression. Will be changed after visualization refactoring.|
|dataFieldsFormats|Optional property, which is needed to display visualization of alert type expression. Will be changed after visualization refactoring.|

## Build and register Action Types

Kibana ships with a set of built-in action types UI:

|Type|Id|Description|
|---|---|---|
|[Server log](#server-log)|`.log`|Logs messages to the Kibana log using `server.log()`|
|[Email](#email)|`.email`|Sends an email using SMTP|
|[Slack](#slack)|`.slack`|Posts a message to a slack channel|
|[Index](#index)|`.index`|Indexes document(s) into Elasticsearch|
|[Webhook](#webhook)|`.webhook`|Send a payload to a web service using HTTP POST or PUT|
|[PagerDuty](#pagerduty)|`.pagerduty`|Trigger, resolve, or acknowlege an incident to a PagerDuty service|

> Every action type should has a server side registration and can has a client side. 

Reviewer 2: ```suggestion
Every Action Type should be registered on the server side and can also be registered on the client side.
```

> Every action type should has a server side registration and can has a client side. 

Reviewer 3: ```suggestion
Every action type should be registered server side, and can be optionally registered client side. 
```

![PagerDuty action form](https://i.imgur.com/xxXmhMK.png)

## Action type model definition

Each action type should be defined as `ActionTypeModel` object with the next properties:
```
  id: string;
  iconClass: string;
  selectMessage: string;
  actionTypeTitle?: string;
  validateConnector: (connector: any) => ValidationResult;
  validateParams: (actionParams: any) => ValidationResult;
  actionConnectorFields: React.FunctionComponent<any> | null;
  actionParamsFields: any;
```
|Property|Description|
|---|---|
|id|Action type id. Should be the same as on server side.|
|iconClass|Icon of action type, that will be displayed on the select card in UI.|
|selectMessage|Short description of action type responsibility, that will be displayed on the select card in UI.|
|validateConnector|Validation function for action connector.|
|validateParams|Validation function for action params.|
|actionConnectorFields|React functional component for building UI of current action type connector.|
|actionParamsFields|React functional component for building UI of current action type params. Displayed as a part of Create Alert flyout.|

## Register action type model

> There are two ways of registration new action type UI:

Reviewer 3: ```suggestion
There are two ways to register a new action type UI:
```

1. Directly in `triggers_actions_ui` plugin. In this case action type will be available in the Alerts and Actions management section.
Registration code for a new action type model should be added to the file `x-pack/plugins/triggers_actions_ui/public/application/components/builtin_action_types/index.ts`
Only registered action types are available in UI.

2. Register action type in other plugin. In this case action type will be available only in the current plugin UI. 
It should be done by importing dependency `TriggersAndActionsUIPublicPluginSetup` and adding the next code on plugin setup:

```
function getSomeNewActionType() {
  return { ... } as ActionTypeModel;
}

triggers_actions_ui.actionTypeRegistry.register(getSomeNewActionType());
```

## Create and register new action type UI

> Before starting the UI implementation, the proper [server side registration](https://github.com/elastic/kibana/tree/master/x-pack/plugins/actions#kibana-actions-configuration) should be done first.

Reviewer 3: ```suggestion
Before starting the UI implementation, the [server side registration](https://github.com/elastic/kibana/tree/master/x-pack/plugins/actions#kibana-actions-configuration) should be done first.
```

import { EuiTextArea } from '@elastic/eui';
import {
  ActionTypeModel,
  ValidationResult,
  ActionConnectorFieldsProps,
  ActionParamsProps,
} from '../../../types';

interface ExampleActionParams {
  message: string;
}

const ExampleParamsFields: React.FunctionComponent<ActionParamsProps<ExampleActionParams>> = ({
  actionParams,
  editAction,
  index,
  errors,
}) => {
  const { message } = actionParams;
  return (
    <Fragment>
      <EuiTextArea
        fullWidth
        isInvalid={errors.message.length > 0 && message !== undefined}
        name="message"
        value={message || ''}
        onChange={e => {
          editAction('message', e.target.value, index);
        }}
        onBlur={() => {
          if (!message) {
            editAction('message', '', index);
          }
        }}
      />
    </Fragment>
  );
};
```

4. Extend registration code with the new action type register in the file `x-pack/plugins/triggers_actions_ui/public/application/components/builtin_action_types/index.ts`
```
import { getActionType as getExampledActionType } from './example';
...

...
actionTypeRegistry.register(getExampledActionType());
```

> After this four steps new `Example Action Type` is available in UI of Create connector:

Reviewer 3: ```suggestion
After these four steps, the new `Example Action Type` is available in UI of Create connector:
```

![Example Action Type is in the select cards list](https://i.imgur.com/PTYdBos.png)

> Click on select cart for `Example Action Type` will open connector form that was created in step 2:

Reviewer 3: ```suggestion
Clicking on the select card for `Example Action Type` will open the connector form that was created in step 2:
```

![Example Action Type connector](https://i.imgur.com/KdxAXAs.png)

Example Action Type is in the select cards list of Create Alert flyout:
![Example Action Type is in the select cards list of Create Alert flyout](https://i.imgur.com/CSRBkFe.png)

> Click on select cart for `Example Action Type` will open action type Add Action form 

Reviewer 2: ```suggestion
Click on `Example Action Type`, which will open the Action Type's _Add Action_ form 
```

> Click on select cart for `Example Action Type` will open action type Add Action form 

Reviewer 3: ```suggestion
Clicking on the select card for `Example Action Type` will open the action type Add Action form 
```
